lib/cloud/gcp/kubernetes.go

```diff
@@ -0,0 +1,270 @@
+	if err := cfg.CheckAndSetDefaults(ctx); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &gkeClient{cfg}, nil
+}
+
+// gkeClient implements the GKEClient interface by wrapping container.ClusterManagerClient
+// and oauth2.TokenSource for interacting with GCP Kubernetes Service.
+type gkeClient struct {
+	GKEClientConfig
+}
+
+// ListClusters lists the GCP GKE clusters that belong to the projectID and are
+// located in location.
+// location supports wildcard "*".
+func (g *gkeClient) ListClusters(ctx context.Context, projectID string, location string) ([]GKECluster, error) {
+	if len(projectID) == 0 {
+		return nil, trace.BadParameter("projectID must be set")
+	}
+	if len(location) == 0 {
+		return nil, trace.BadParameter("zone must be set")
+	}
+
+	location = convertLocationToGCP(location)
+
+	res, err := g.GKEClientConfig.ClusterClient.ListClusters(
+		ctx,
+		&containerpb.ListClustersRequest{
+			Parent: fmt.Sprintf("projects/%s/locations/%s", projectID, location),
+		},
+	)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	var clusters []GKECluster
+	for _, cluster := range res.Clusters {
+		clusters = append(clusters, GKECluster{
+			Name:        cluster.Name,
+			Description: cluster.Description,
+			ProjectID:   projectID,
+			Labels:      cluster.ResourceLabels,
+			Status:      cluster.Status,
+			Location:    cluster.Location,
+		})
+	}
+
+	return clusters, nil
+}
+
+// GetClusterRestConfig returns the Kubernetes client config to connect to the
+// specified cluster. The access token is based on the default credentials configured
+// for the current GCP Service Account and must include the following permissions:
+// - container.clusters.get
+// - container.clusters.impersonate
+// - container.clusters.list
+// - container.pods.get
+// - container.selfSubjectAccessReviews.create
+// - container.selfSubjectRulesReviews.create
+// It also returns the token expiration time from which the token is no longer valid.
+func (g *gkeClient) GetClusterRestConfig(ctx context.Context, cfg ClusterDetails) (*rest.Config, time.Time, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, time.Time{}, trace.Wrap(err)
+	}
+
+	// Get cluster from cloud to extract the CA certificate.
+	res, err := g.GKEClientConfig.ClusterClient.GetCluster(
+		ctx,
+		&containerpb.GetClusterRequest{
+			Name: cfg.toGCPEndpointName(),
+		},
+	)
+	if err != nil {
+		return nil, time.Time{}, trace.Wrap(err)
+	}
+
+	// Generate a SA Authentication Token.
+	token, err := g.GKEClientConfig.TokenSource.Token()
+	if err != nil {
+		return nil, time.Time{}, trace.Wrap(err)
+	}
+
+	restCfg, err := getTLSConfig(res, token.AccessToken)
+
+	return restCfg, token.Expiry, trace.Wrap(err)
+}
+
+// convertLocationToGCP checks if the location is a Teleport wildcard `*` and
+// replaces it with the GCP wildcard, otherwise returns the location.
+func convertLocationToGCP(location string) string {
+	if location == types.Wildcard {
```

Comment:
Wasn't there a check above that "location" can't be wildcard?

Reply:
The check only applies when you want to describe a single cluster. 
`projects/{project_id}/locations/{location}/clusters/{clusterID}` - in this case {location} cannot be a wildcard.

When you want to receive all available clusters, you can use wildcards.
`projects/{project_id}/locations/{location}` - in this case {location} can be a wildcard